File: cgi-bin/flagged.py
#!/usr/bin/python 
# -*- coding: utf-8  -*-

import time, sys
import MySQLdb
db = MySQLdb.connect(read_default_file="/home/hroest/.my.cnf")
sys.path.append( '/home/hroest/' )
import replag_lib, db_api
import cgi 
form = cgi.FieldStorage()   # FieldStorage object to
start = time.time()
print "Content-type: text/html; charset=utf-8"
print ""

class Page:
    def __init__(self): pass

def get_time_diff(timestamp, return_as = 'D'):
    import datetime
    now = datetime.datetime.now()

    year = now.year     - int(timestamp[:4])
    mon =  now.month    - int(timestamp[4:6])
    day =  now.day      - int(timestamp[6:8])
    hour = now.hour     - int(timestamp[8:10])
    min =  now.minute   - int(timestamp[10:12])

    if return_as == 'D':
        days = 356*year + 30*mon + day
        return days

def db_get_unreviewed( language, category, c):

    category = db_api.make_db_safe( category )
    sql = """SELECT page_id,page_title,page_latest,fp_stable, rev_len
    FROM page,categorylinks,flaggedpages, revision
    WHERE cl_to="%s" AND cl_from=page_id AND fp_page_id=page_id 
    AND page_latest<>fp_stable AND page_namespace=0
    AND rev_id=page_latest""" % category;

    db = language + 'wiki_p' ;
    tmp = c.execute( sql )
    res = c.fetchall()

    reviewed_pages = []
    for o in res:
        page = Page()
        page.id          = o[0]
        page.title       = o[1]
        page.current_rev = o[2]
        page.stable_rev  = o[3]
        page.latest_size = o[4]

        #revision[0] is the latest flagged revision
        #revision[1] is the first unflagged revision 
        #we do not store the revisions for memory reasons
        sql = """select rev_id,  rev_comment, rev_user_text, 
        rev_timestamp, rev_len
        from revision where rev_page = %s
        and rev_id >= %s""" % (page.id, page.stable_rev)
        c.execute( sql ) 
        revisions = c.fetchall()

        stable_len = revisions[0][4]
        page.size_diff = page.latest_size - stable_len

        page.first_unflagged_timestamp = revisions[1][3]
        reviewed_pages.append(page)

    return reviewed_pages;


if not form.has_key('category'):
    print """ <h1>Flagged Revisions Tool</h1>
    <i>%(title)s</i><br/>
    <form method='get' action='flagged.py'>
    <table>
    <tr><th>%(language_title)s</th><td><input type='text' 
        name='language' value='%(language)s' size='30' /></td></tr>
    <tr><th>%(catname)s</th><td><input type='text' name='category' value='' 
        size='30' /></td></tr>
    <tr><th>exclude*</th><th>
        <textarea name='exclude' rows='3' cols='80'></textarea></th>
    </tr>
    <tr><th>%(depth_title)s**</th><td><input type='text' name='depth' 
        value='%(depth)s' /> </td></tr>
    <tr><th></th><td><input type='submit' name='doit' value='%(go)s' /></td></tr>
    <tr><th>%(sortby)s</th><th>
       <Input type = 'Radio' Name='sortby' value= 'size' checked='checked' >Size
       <Input type = 'Radio' Name='sortby' value= 'size_reverse'>Size (reverse)
    </tr>
    <tr><th></th><th>
       <Input type = 'Radio' Name='sortby' value= 'title'>Title
       <Input type = 'Radio' Name='sortby' value='title_reverse'>Title (reverse)
    </tr>
    <tr><th></th><th>
       <Input type = 'Radio' Name='sortby' value= 'time'>Time
       <Input type = 'Radio' Name='sortby' value= 'time_reverse'>Time (reverse)
    </tr>
    </table>
    <br/>
    *Format: \"cat1;cat2;cat3\".  <br/>
    This might not do what you expect it to do.<br/>
    If you list a category under \"exclude\", this category is not considered<br/>
    in the recursive search for articles. If the article you want to exclude is<br/>
    encountered in a different category during the search, it will still be <br/>
    included.
    <br/>
    <br/>
    ** 1=%(expl1)s, -99 = %(expl2)s
    </form>
    <p> <a href='http://toolserver.org/~hroest/'>Zurueck zur Uebersicht</a> </p>
    """ % { 'go' : 'Los!', 'catname' : 'Kategorie', 'depth' : 'Tiefe', 
        'depth_title' : 'Tiefe', 'language_title' : 'Sprache',
        'depth' : 1, 'sortby' : 'Sortiere nach', 'title' :  
        "Zeigt Artikel in (Unter)kategorien, die nachgesichtet werden müssen.", 
        'expl1' : 'Kategorie und direkte Unterkategorien', 'language' : 'de',
        'expl2' :'Alle Unterkategorien'  }
    exit()

#///////////////////////////////////////////////////////////////////////////
#//   OPTIONS
#///////////////////////////////////////////////////////////////////////////


language        =     form.getvalue( 'language')
category        =     form.getvalue( 'category')
depth           = int(form.getvalue( 'depth') )
sortby          =     form.getvalue( 'sortby')
exclude_noncode =     form.getvalue( 'exclude')
show_color_at   =     300
diff_only       =     0
exclude = exclude_noncode.replace( " ", "_")
exclude = exclude.split(";")

print "Excluding: ";
for ex in exclude: print ex + ";"
print "<br/>";

start = time.time()
c = db.cursor()
data = db_api.db_get_articles_in_category ( language, category, c, 
                                       depth-1, 14, exclude ) 
if not category in data: data.append( category )
print "Durchsuche " + str( len ( data ) ) + " Kategorien.";

output = []
for d in data:
    my_unreviewed = db_get_unreviewed ( language, d, c);
    for unrev in my_unreviewed:
        if not unrev.title in [o.title for o in output]: output.append( unrev)

print len( output )

#now we have all unreviewed pages, lets sort them
if (sortby == 'size')           :
    output.sort( lambda x, y: cmp(x.size_diff, y.size_diff)  )
elif (sortby == 'title')        :
    output.sort( lambda x, y: cmp(x.title, y.title)  )
elif (sortby == 'time')         :
    output.sort( lambda x, y: -cmp(x.first_unflagged_timestamp, 
                                  y.first_unflagged_timestamp)  )
elif (sortby == 'size_reverse') :
    output.sort( lambda x, y: -cmp(x.size_diff, y.size_diff)  )
elif (sortby == 'title_reverse'):
    output.sort( lambda x, y: -cmp(x.title, y.title)  )
elif (sortby == 'time_reverse') :
    output.sort( lambda x, y: cmp(x.first_unflagged_timestamp, 
                                  y.first_unflagged_timestamp)  )


printlist = '';
taburls = []
for page in output:
    t =  page.title.replace( "_" , " " )
    old_id = page.stable_rev
    url = "http://de.wikipedia.org/w/index.php?title=" + t 
    url += "&diffonly=0&oldid=%s&diff=cur" % old_id

    timediff = get_time_diff(page.first_unflagged_timestamp)
    diff = page.size_diff
    if abs ( diff ) > show_color_at:bg = '#AAAAAA';
    else: bg = '#FFFFFF';
    if diff > 0:
        diff_display="<span style='color:green;background:$bg'>+%s</span>"%diff
    elif diff < 0:
        diff_display = "<span style='color:red;background:$bg'>%s</span>"%diff
    else: diff_display = "<span style='background:$bg'>%s</span>"%diff

    note = "Diff = " + diff_display
    note += " Alter = "  + str(timediff) + " Tage" ;

    printlist += '<li><a target="_blank" href="%s">' % url
    printlist += t + "</a>%s</li>\n" % note;

    taburls.append(url)

print "<ol>" + printlist + "</ol>";
print "\n\n<br/>Today: " +  "\n<br/>";
end = time.time()
timediff = end - start
print "This query took %s seconds. <br/>" % timediff

print str(len( output ) ) + " nachzusichtende Artikel gefunden. "


#********************************************* 
#Javascript from Magnus
#********************************************* 
#
out = ''
print "<script type='text/javascript'>"

if len( taburls ) > 10:
    print "function open10 () {\n"
    for a in range(10):
        print "window.open ('"+taburls[a]+"', '_blank');" ;
    print "return false;};\n\n" ;
    out += "<a href='#' onclick='open10()'>Open first 10 in tabs</a> " ;

if len( taburls ) > 20:
    print "function open20 () {\n"
    for a in range(20):
        print "window.open ('"+taburls[a]+"', '_blank');" ;
    print "return false;};\n\n" ;
    out += "<a href='#' onclick='open20()'>Open first 20 in tabs</a> " ;

if len( taburls ) > 50:
    print "function open50 () {\n"
    for a in range(50):
        print "window.open ('"+taburls[a]+"', '_blank');" ;
    print "return false;};\n\n" ;
    out += "<a href='#' onclick='open50()'>Open first 50 in tabs</a> " ;

print "function open_all() {\n" ;
for tab in taburls:
    print "window.open ('"+tab+"', '_blank');\n" ;
print "return false;};\n\n" ;
out += "<a href='#' onclick='open_all()'>" ;
out += "Open all in tabs" ;
out += "</a> " ;

print "</script>" 
print out

print '<p> <a href="http://toolserver.org/~hroest/">Zurueck zur Uebersicht</a> </p>';
print '</html>' ;

File: flagged.php
<?PHP

/*
 * 
 * Deep out of Sight IMproved
 * (c) by Hannes Röst unless otherwise indicated.
 * this code is adapted from magnus' deep_out_of_sight.php 
 * Released under GPL.
 *
 */
include_once ( "db_library.php");
include_once ( "tools_library.php");
$mb = 30;
ini_set ('memory_limit', 1024*1024*$mb);

function get_fr( $language, $user ) {
    if ( !isset ( $mysql_con ) ) {$mysql_con = set_up_db($language);}
    $sql = "SELECT count(*) FROM flaggedrevs, user WHERE
            fr_user = user_id AND user_name = \"$user\"";

    $db = $language . 'wiki_p' ;
    $res = mysql_db_query ( $db , $sql , $mysql_con ) ;

}

//this will get all unreviewed pages that are in a given category
function db_get_unreviewed( $language, $category, $fast = 99, $namespace=0) {
    //global $mysql_con;
    if ( !isset ( $mysql_con ) ) {$mysql_con = set_up_db($language);}
    make_db_safe ( $category ) ;

    $sql = "SELECT page_id,page_title,page_latest,fp_stable, rev_len
    FROM page,categorylinks,flaggedpages, revision
    WHERE cl_to=\"$category\" AND cl_from=page_id AND fp_page_id=page_id 
    AND page_latest<>fp_stable AND page_namespace=$namespace
    AND rev_id=page_latest";

    $db = $language . 'wiki_p' ;
    $res = mysql_db_query ( $db , $sql , $mysql_con ) ;

    $reviewed_pages = array () ;
    while ( $o = mysql_fetch_object ( $res ) ) {
        $latest_size = $o ->rev_len;
        $current_rev = $o->page_latest;
        $stable_rev = $o->fp_stable;

        if($stable_rev >= $current_rev) 
        {
            // something is wrong here, lets leave
            print "<br/>Something is wrong here with ";
            print  $o->page_title;
            print " <br/>";
            continue;
        }

        if($fast == 1) {
            $reviewed_pages[ $o-> page_id ] = array( 'title' => $o->page_title,
                'stable_id' => $stable_rev);continue;}

        //now get the len of the last stable
        $sql = "SELECT rev_len FROM revision WHERE rev_id=$stable_rev";
        $res_inner = mysql_db_query ( $db , $sql , $mysql_con ) ;
        $o3 = mysql_fetch_object ( $res_inner ) ;
        $size_diff = $latest_size - $o3 -> rev_len;

        if($fast == 2) {
            $reviewed_pages[ $o-> page_id ] = array( 'title' => $o->page_title,
                 'diff' => $size_diff, 'stable_id' => $stable_rev);continue;}

        //now get the first unflagged revision
        while ( $current_rev != $stable_rev) {
            //$before_current_rev = $current_rev;
            $sql = "SELECT rev_parent_id, rev_timestamp FROM revision 
                    WHERE rev_id=$current_rev";
            $res_inner = mysql_db_query ( $db , $sql , $mysql_con ) ;
            $o2 = mysql_fetch_object ( $res_inner ) ;
            $current_rev = $o2->rev_parent_id;
            if($current_rev == 0) {break;}
            if(!$o2) {break;}
        }
        $first_unflagged_timestamp = $o2->rev_timestamp ;


        $reviewed_pages[ $o-> page_id ] = array( 'title' => $o->page_title, 
            'timestamp' => $first_unflagged_timestamp, 'diff' => $size_diff ,
            'stable_id' => $stable_rev);
     }
    return $reviewed_pages;
}

///////////////////////////////////////////////////////////////////////////
//   OPTIONS
///////////////////////////////////////////////////////////////////////////
$language = get_request ( 'language' , 'de' ) ;
$category = get_request ( 'category' , '' ) ;
$depth = get_request ( 'depth' , 1 ) ;
$namespace = get_request ( 'namespace' , 0 ) ;
$sortby = get_request ( 'sortby' , 'category' ) ;
$exclude_noncode = get_request ( 'exclude' , '' ) ;
$fast = get_request ( 'fast' , 1 ) ;
$show_color_at = get_request( 'show_color_at', 300);
$diff_only = get_request( 'diff_only', 0);

//we dont need these options
$extended = get_request ( 'extended' , 0 ) ;
$simple = get_request ( 'simple' , 0 ) ;

//rss?
$mode = get_request ( 'mode' , '' ) ;
$catname = 'Kategorie' ;


//print $exclude[0];
//print $exclude[1];
//$exclude = array( "Dendrologe", "Botaniker", "Forstwissenschaftler" );

print '<html><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"></head>' ;
if ($category == '') {
        $t1 = 'Zeigt Artikel in (Unter)kategorien, die nachgesichtet werden müssen.' ; 
        $t2 = 'Sprache' ;
        $t3 = 'Tiefe' ;
        $t4 = 'Kategorie und direkte Unterkategorien' ;
        $t4B = 'Alle Unterkategorien' ;
        $t5= 'Sortiere nach' ;
        $t6= 'Namensraum' ;
        $t9 = 'Los!' ;

    print "<body>
    <h1>Flagged Revisions Tool</h1>
    <i>$t1</i><br/>
    <form method='get' action='flagged.php'>
    <table>
    <tr><th>$t2</th><td><input type='text' name='language' value='$language' size='30' /></td></tr>
    <tr><th>$catname</th><td><input type='text' name='category' value='' size='30' /></td></tr>
    <tr><th>exclude*</th><th>
        <textarea name='exclude' rows='3' cols='80'></textarea></th>
    </tr>
    <tr><th>$t3**</th><td><input type='text' name='depth' value='$depth' /> </td></tr>
    <tr><th>$t6</th><td><input type='text' name='namespace' value='$namespace' /> 0=articles, 14=categories</td></tr>
    <tr><th></th><td><input type='submit' name='doit' value='$t9' /></td></tr>
    <tr><th>$t5</th><th>
       <Input type = 'Radio' Name ='sortby' value= 'size' checked='checked' >Size
       <Input type = 'Radio' Name ='sortby' value= 'size_reverse'  >Size (reverse)
    </tr>
    <tr><th></th><th>
       <Input type = 'Radio' Name ='sortby' value= 'title'         >Title
       <Input type = 'Radio' Name ='sortby' value= 'title_reverse' >Title (reverse)
    </tr>
    <tr><th></th><th>
       <Input type = 'Radio' Name ='sortby' value= 'time'          >Time
       <Input type = 'Radio' Name ='sortby' value= 'time_reverse'  >Time (reverse)
    </tr>
    </table>
    <br/>
    *Format: \"cat1;cat2;cat3\".  <br/>
    This might not do what you expect it to do.<br/>
    If you list a category under \"exclude\", this category is not considered<br/>
    in the recursive search for articles. If the article you want to exclude is<br/>
    encountered in a different category during the search, it will still be <br/>
    included.
    <br/>
    <br/>
    ** 1=$t4, -99 = $t4B
    </form>
    <p> <a href='http://toolserver.org/~hroest/'>Zurueck zur Uebersicht</a> </p>
    </body>" ;
    print '</html>' ;
    exit  ;

}

// Process input
$exclude = str_replace( " ", "_", $exclude_noncode);
$exclude = split(";", $exclude);

//print $exclude;
print "Excluding: ";
foreach($exclude AS $ex) {
    print mydecode($ex) . ";";
}
print "<br/>";

//some options are mutually exclusive -- you cannot be fast and sort by time
if ($fast == 1) {
    if ($sortby =='title' or $sortby == 'title_reverse' or $sortby == 'category'){
        $myfast = 1;}  
    if ($sortby == 'size' or $sortby == 'size_reverse'){$myfast = 2;}
    if ($sortby == 'time' or $sortby == 'time_reverse'){$myfast = 99;}
} else $myfast = 99;


//Lets start!
//We get the categories we need
$starttime = microtime(true); 

//get all articles in category with ns == 14 (category ns)
$data = db_get_articles_in_category ( $language , $category , $depth-1 , 14 , 
    $exclude ) ;

if ( count ( $data ) == 0 ) {
    $category = utf8_encode ( $category ) ;
    $data = db_get_articles_in_category( $language , $category , $depth-1 , 14 ,
    $exclude ) ;
}
//now also add the root category itself
if ( !isset ( $data[$category] ) ) array_unshift ( $data , $category ) ;

print "Durchsuche " .  count ( $data ) . " Kategorien.";

$output = array () ;
foreach ( $data AS $d ) {
    $my_unreviewed = db_get_unreviewed ( $language, $d, $myfast, $namespace);
    foreach ( $my_unreviewed AS $k => $v ) { // Show each article only once
        if ( isset ( $output[$k] ) ) {unset ( $my_unreviewed[$k] );}
    }

    //print mydecode($d) . "\n <br/>";
    //dont need this
    //if ( count ( $my_unreviewed ) == 0 ) continue ;

    //do NOT use array_merge, this will rename keys!
    $output =  $output + $my_unreviewed;
}

if ($sortby == 'size')              array_sort($output,'diff'); 
else if ($sortby == 'title')        array_sort($output,'title'); 
else if ($sortby == 'time')         array_sort($output,'!timestamp'); 
else if ($sortby == 'size_reverse') array_sort($output,'!diff'); 
else if ($sortby == 'title_reverse')array_sort($output,'title'); 
else if ($sortby == 'time_reverse') array_sort($output,'timestamp'); 
else if ($sortby == 'category') {;}

$printlist = '';
$taburls = array () ;
foreach ($output AS $page) {
    $t =  str_replace ( "_" , " " , $page['title'] );
    $old_id = $page['stable_id'];
    $url = get_wikipedia_url ( $language , $t ) . 
        "&diffonly=$diff_only&oldid=$old_id&diff=cur" ;

    $timediff = get_time_diff( $page['timestamp'], 'D' );
   
    //size
    $diff = $page['diff'];
    if ( abs ( $diff ) > $show_color_at) $bg = '#AAAAAA';
    else $bg = '#FFFFFF';
    if ( $diff > 0 ) 
        $diff_display = "<span style='color:green;background:$bg'>+$diff</span>" ;
    else if ( $diff < 0 ) 
        $diff_display = "<span style='color:red;background:$bg'>$diff</span>" ;
    else $diff_display = "<span style='background:$bg'>$diff</span>" ;

    $note = " ";
    if($myfast > 1)  $note .= "Diff = " . $diff_display;
    if($myfast > 2)  $note .= " Alter = "  . $timediff . " Tage" ;

    $printlist .= '<li><a target="$target" href="' . $url . '">' ;
    $printlist .= $t . "</a>$note</li>\n" ;

    $taburls[] = $url ;
}

///////////////////////////////////////////////////////////////////////////
// Print out starts here
///////////////////////////////////////////////////////////////////////////

flush();

print "<ol>" . $printlist . "</ol>";
print "\n\n<br/>Today: " . date(DATE_RSS) . "\n<br/>";
$endtime = microtime(true); 
$timediff = $endtime - $starttime;
print "This query took $timediff seconds. <br/>";

print count( $output ) . " nachzusichtende Artikel gefunden. ";


//
//********************************************* 
//Javascript from Magnus
//********************************************* 
//
$out = '' ;
print "<script type='text/javascript'>" ;

if ( count ( $taburls ) > 20 ) {
    print "function open20 () {\n" ;
    for ( $a = 0 ; $a < 20 ; $a++ ) {
        print "window.open ('".$taburls[$a]."', '_blank');\n" ;

    }
    print "return false;};\n\n" ;
    $out .= "<a href='#' onclick='open20()'>Open first 20 in tabs</a> " ;
}

if ( count ( $taburls ) > 50 ) {
    print "function open50 () {\n" ;
    for ( $a = 0 ; $a < 50 ; $a++ ) {
        print "window.open ('".$taburls[$a]."', '_blank');\n" ;

    }
    print "return false;};\n\n" ;
    $out .= "<a href='#' onclick='open50()'>Open first 50 in tabs</a> " ;
}

print "function open_all() {\n" ;
for ( $a = 0 ; $a < count($taburls) ; $a++ ) {
    print "window.open ('".$taburls[$a]."', '_blank');\n" ;
}
print "return false;};\n\n" ;
$out .= "<a href='#' onclick='open_all()'>" ;
$out .= "Open all in tabs" ;
$out .= "</a> " ;

print "</script>$out" ;

print '<p> <a href="http://toolserver.org/~hroest/">Zurueck zur Uebersicht</a> </p>';
print '</html>' ;

?>
